lib/spack/spack/schema/mirrors.py

```diff
@@ -10,19 +10,60 @@
 """
 from typing import Any, Dict
 
+
+string_or_variable = {
+    "oneOf": [
+        {
+            "type": "object",
+            "required": ["variable"],
+            "additionalProperties": False,
+            "properties": {"variable": {"type": "string"}},
+        },
+        {"type": "string"},
+    ]
+}
+
+
 #: Common properties for connection specification
 connection = {
     "url": {"type": "string"},
     # todo: replace this with named keys "username" / "password" or "id" / "secret"
     "access_pair": {
-        "type": "array",
-        "items": {"type": ["string", "null"], "minItems": 2, "maxItems": 2},
+        "oneOf": [
+            {
+                "type": "array",
+                "items": {"minItems": 2, "maxItems": 2, **string_or_variable},
+            },  # deprecated
+            {
```

Comment:
This isn't strictly required, but I think it makes sense as a pair is two things and being able to clearly list them in the config is useful.

I do not have an option for just listing the `secret` in the new object format.